Validate port-options

amadeuszpawlik opened the pull request:
Validate `port`-options, so that invalid values are rejected early in the startup.
Ports are `uint16_t`s, which effectively limits a port's value to <=65535. As discussed in https://github.com/bitcoin/bitcoin/pull/24116 and https://github.com/bitcoin/bitcoin/pull/24344, port "0" is considered invalid too.
Proposed in https://github.com/bitcoin/bitcoin/issues/21893#issuecomment-835784223

The `SplitHostPort(std::string in, uint16_t& portOut, std::string& hostOut)` now returns a bool that indicates whether the port value was set and within the allowed range. This is an improvement that can be used not only for port validation of options at startup, but also in rpc calls, etc,

amadeuszpawlik committed:
Validate port value in `SplitHostPort`

Forward the validation of the port from `ParseUInt16(...)`.
Consider port 0 as invalid.
Add suitable test for the `SplitHostPort` function.
Add doxygen description to the `SplitHostPort` function.

amadeuszpawlik committed:
Validate `port` options

Check `port` options for invalid values (ports are parsed as uint16, so
in practice values >65535 are invalid; port 0 is undefined and therefore
considered invalid too). This allows for an early rejection of faulty
values and an supplying an informative message to the user.

Splits tests in `feature_proxy.py` to cover both invalid `hostname`
and `port` values.

Adds a release-note as previously valid `-port` and `-rpcport` values
can now result in errors.

jonatack commented:
I'd suggest searching for and looking at a previous pull request proposal to do this. Consider also unit testing and Doxygen docs.

amadeuszpawlik commented:
Added doxygen description
Added/updated test for `SplitHostPort`
@jonatack could you clarify 
>Consider also the various places in the codebase where users can input port numbers.
Have I missed something, like a config?

klementtan reviewed (commented):
You might wanna add the validation at [`addpeeraddress`](https://github.com/bitcoin/bitcoin/blob/master/src/rpc/net.cpp#L935) too. Currently, the port number will overflow if it is invalid.

amadeuszpawlik commented:
Thanks for your input, @klementtan !
I addressed your comment in a1c591453fbd3f28772a230720675f4e9ff74978

klementtan commented on src/util/strencodings.cpp:
```diff
@@ -128,6 +127,25 @@ void SplitHostPort(std::string in, uint16_t& portOut, std::string& hostOut)
+    SplitHostPort(in, portStr, hostOut);
+
+    uint16_t n;
+    if (!portStr.empty() && ParseUInt16(portStr, &n))
```
The value `portOut` will be undefined behavior if `portStr = "-1"` as `ParseUInt16` will return false before assigning a value to `n`

amadeuszpawlik replied:
I don't know if I agree with you on this point:
My intention was not to change the behaviour of `SplitHostPort(...)` but rather to gain access to the port portion of a socket address as a string.
If you for example take a look at [src/httpserver.cpp](https://github.com/bitcoin/bitcoin/blob/a748782a11f0b23f8dcdf167f45d68c863524887/src/httpserver.cpp#L307&L308&L309) in `HTTPBindAddresses(...)` this function is used as follows:
`uint16_t port{some_default_port};
SplitHostPort(..., port, ...);`
The same can be seen in `src/bitcoin-cli.cpp` and `src/net.cpp`: the callers already set default port value before calling this function.

klementtan replied:
That makes sense thanks!

klementtan reviewed (commented):
Thanks for updating it!

I think you might have missed out the scenario where the user enters `-1` as the port. On a1c591453fbd3f28772a230720675f4e9ff74978
```bash
$ ./src/bitcoin-cli -signet addpeeraddress "1.1.1.3" -1
{
  "success": true
}
```

You might wanna refactor the logic in `strencodings.cpp` so that the port validation code could be reused in other places(ie `addpeeraddress`)

amadeuszpawlik commented:
@klementtan 
> think you might have missed out the scenario where the user enters -1 as the port.

You are absolutely right, fixed that. I also added a couple negative port numbers in the tests, thanks.

laanwj commented:
Concept ACK. I would personally call this "validate" (as in input validation), not "santize". In the source base we use "sanitize" only in the context of removing invalid characters from strings in logging.

laanwj commented on src/util/strencodings.h:
```diff
@@ -65,7 +65,32 @@ std::string EncodeBase32(Span<const unsigned char> input, bool pad = true);
+ * @param[out] portOut  Port-portion of the input, if found and parsable.
+ * @param[out] hostOut  Host-portion of the input, if found.
+ */
 void SplitHostPort(std::string in, uint16_t& portOut, std::string& hostOut);
```
I would prefer this to be more explicit about parse errors. For example:

- Return a boolean `true` if valid `false` otherwise
- Return `std::optional<uint16_t>` (or even `std:::optional<std::pair<string, uint16_t>>`)

Either is a lot clearer than "we don't assign anything if the port is invalid", which is somewhat indirect behavior.

amadeuszpawlik replied:
Changed to return bool

prayank23 reviewed (changes requested):
Concept ACK

Tested by fuzzing `addpeeraddress`, did not try config params: `-port -rpcport -bind -i2psam -onion-proxy -rpcbind -torcontrol -whitebind -zmq*`

Example of a response for invalid port number or you get parse error:

```
HTTP/1.1 500 Internal Server Error

Content-Type: application/json

Date: Thu, 02 Dec 2021 21:44:11 GMT

Content-Length: 90

Connection: close



{"result":null,"error":{"code":-8,"message":"Invalid port specified: 89614"},"id":"test"}

```

:x: One invalid user input returned 200 response. Negative zero from this [wordlist](https://raw.githubusercontent.com/danielmiessler/SecLists/master/Fuzzing/big-list-of-naughty-strings.txt)

Request:

```
POST / HTTP/1.1

Host: 127.0.0.1:18662

Authorization: Basic dXNlcjY6cGFzczY=

Content-Type: text/plain

Content-Length: 92

Connection: close



{"jsonrpc": "1.0", "id": "test", "method": "addpeeraddress", "params": ["4.4.4.4",-0, true]}

```

Response:

```
HTTP/1.1 200 OK

Content-Type: application/json

Date: Thu, 02 Dec 2021 22:15:28 GMT

Content-Length: 54

Connection: close



{"result":{"success":true},"error":null,"id":"test"}

```

amadeuszpawlik commented:
@prayank23 perhaps I should keep this PR to options only, and take a look at RPCs in a separate one?

amadeuszpawlik commented:
I reconsidered the `addpeeraddress` after @prayank23's [comment](https://github.com/bitcoin/bitcoin/pull/22087#pullrequestreview-822083480) and decided to remove that case from this PR. Validating port values in RPCs does not contribute to "early rejection of invalid port values at startup". Please correct me if I'm wrong.

Apropos that: looking at the `-0` case in `addpeeraddress` I came to think that the best way to solve this would be to add a helper function `get_uint16()` alongside [`get_int()`](https://github.com/bitcoin/bitcoin/blob/ea989deecc8d4d02bb401c0ac7cd7a0543e3454c/src/univalue/include/univalue.h#L173) in [`univalue_get.cpp`](https://github.com/bitcoin/bitcoin/blob/master/src/univalue/lib/univalue_get.cpp), because the way it is now, the nonsensical `-` in `-0` gets dealt with before I can get hold of it in `rpc.cpp`; would it make sense to do that or should I just ignore this case as -0 == 0 anyhow?

luke-jr reviewed (changes requested):
Please put the error string in one place, with a substitution for the option name. This helps translators only need to translate it once, and avoids accidental translation of the option name.

thonkle commented on src/init.cpp:
```diff
@@ -1276,6 +1276,68 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+    if (args.IsArgSet("-port")) {
+        uint64_t portIn = args.GetIntArg("-port", 0);
+        if (portIn > std::numeric_limits<uint16_t>::max())
+            return InitError(strprintf(_("Invalid port specified in -port: '%d'"), portIn));
```
This should also error for port 0.

See discussions in https://github.com/bitcoin/bitcoin/pull/24116 and https://github.com/bitcoin/bitcoin/pull/24344

luke-jr replied:
And for less than 0... (change type above to `int64_t`)

amadeuszpawlik replied:
Thanks for the input, guys. I applied your suggestions + set `bool SplitHostPort(...)` to consider port 0 as invalid.

luke-jr commented on src/init.cpp:
```diff
@@ -1276,6 +1276,69 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
     RegisterValidationInterface(node.peerman.get());
 
+    // Check port numbers
+    const bilingual_str port_error = _("Invalid port specified in ");
```
```suggestion
    const bilingual_str port_error = _("Invalid port specified in %s: %s");
```

Other languages may need to change the ordering of parameters.

amadeuszpawlik replied:
Yeah, I did that initially but the linter complains, any tips?

luke-jr replied:
Check out the functions at the bottom of `src/util/error.cpp`

amadeuszpawlik replied:
Thanks for the review Luke, much appreciated

luke-jr commented on src/init.cpp:
```diff
@@ -1276,6 +1276,69 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+
+    std::string hostOut;
+    uint16_t portOut{0};
+    for (const std::string& socketAddr : args.GetArgs("-bind")) {
```
A lot of repetition. Maybe an outer for loop over each option name?

amadeuszpawlik replied:
Done, good point

luke-jr commented on src/init.cpp:
```diff
@@ -1276,6 +1276,30 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
 
+    // Check port numbers
+    const bilingual_str port_error = _("Invalid port specified in %s: %s");
+    const std::string port_options_int[] = {"-port", "-rpcport"};
+    for (const std::string& port_option : port_options_int) {
```
nit

```suggestion
    for (const std::string& port_option : {
        "-port",
         "-rpcport",
    }) {
```

luke-jr commented on src/init.cpp:
```diff
@@ -1276,6 +1276,30 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+        if (args.IsArgSet(port_option)) {
+            int64_t portIn = args.GetIntArg(port_option, 0);
+            if (portIn <= 0 || portIn > std::numeric_limits<uint16_t>::max())
+                return InitError(strprintf(port_error, port_option, ToString(portIn)));
```
nit: I don't think this needs ToString (strprintf should handle int64_t as `%s` just fine)

```suggestion
                return InitError(strprintf(port_error, port_option, portIn));
```

luke-jr commented on src/init.cpp:
```diff
@@ -1276,6 +1276,30 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+        }
+    }
+
+    std::string hostOut;
+    uint16_t portOut{0};
```
nit: Prefer scoping these just before `SplitHostPort` call

luke-jr commented on src/init.cpp:
```diff
@@ -1276,6 +1276,30 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+
+    std::string hostOut;
+    uint16_t portOut{0};
+    const std::string port_options_str[] = {"-bind", "-i2psam", "-onion", "-proxy", "-rpcbind",
+      "-torcontrol", "-whitebind", "-zmqpubhashblock","-zmqpubhashtx","-zmqpubrawblock", "-zmqpubsequence"};
+    for (const std::string& port_option : port_options_str) {
```
nit

```suggestion
    for (const std::string& port_option : {
        "-bind",
        "-i2psam",
        "-onion",
        "-proxy",
        "-rpcbind",
        "-torcontrol",
        "-whitebind",
        "-zmqpubhashblock",
        "-zmqpubhashtx",
        "-zmqpubrawblock",
        "-zmqpubsequence",
    }) {
```

luke-jr commented on src/init.cpp:
```diff
@@ -1276,6 +1276,30 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+      "-torcontrol", "-whitebind", "-zmqpubhashblock","-zmqpubhashtx","-zmqpubrawblock", "-zmqpubsequence"};
+    for (const std::string& port_option : port_options_str) {
+        for (std::string socketAddr : args.GetArgs(port_option)) {
+            if (port_option == "-bind")
+                socketAddr = socketAddr.substr(0, socketAddr.rfind('='));
```
Include braces, even if it's just a one line body.

```suggestion
            if (port_option == "-bind") {
                socketAddr = socketAddr.substr(0, socketAddr.rfind('='));
            }
```

luke-jr commented on src/init.cpp:
```diff
@@ -1276,6 +1276,30 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+        for (std::string socketAddr : args.GetArgs(port_option)) {
+            if (port_option == "-bind")
+                socketAddr = socketAddr.substr(0, socketAddr.rfind('='));
+            if (!SplitHostPort(socketAddr, portOut, hostOut))
+                return InitError(strprintf(port_error, port_option, socketAddr));
```
braces again

```suggestion
            if (!SplitHostPort(socketAddr, portOut, hostOut)) {
                return InitError(strprintf(port_error, port_option, socketAddr));
            }
```

luke-jr commented on src/util/strencodings.cpp:
```diff
@@ -128,6 +127,22 @@ void SplitHostPort(std::string in, uint16_t& portOut, std::string& hostOut)
+      valid = true;
+    } else if (ParseUInt16(portStr, &n)) {
+      portOut = n;
+      valid = portOut == 0 ? false : true;
```
```suggestion
      valid = (portOut != 0);
```

luke-jr commented on src/util/strencodings.cpp:
```diff
@@ -107,18 +107,17 @@ std::vector<unsigned char> ParseHex(const std::string& str)
 }
 
-void SplitHostPort(std::string in, uint16_t& portOut, std::string& hostOut)
+void SplitHostPort(std::string in, std::string& portOut, std::string& hostOut)
```
I don't think we gain anything by having a string-only parser?

Doing this means `SplitHostPort` will only conditionally validate port range.

luke-jr commented on src/util/strencodings.cpp:
```diff
@@ -107,25 +107,35 @@ std::vector<unsigned char> ParseHex(const std::string& str)
+            }
             in = in.substr(0, colon);
-            portOut = n;
+        } else {
+          valid = true;
```
This is weird. Ending a host/port string with a bare colon should be treated as invalid IMO.

luke-jr commented on src/util/strencodings.h:
```diff
@@ -86,7 +86,16 @@ std::string EncodeBase32(Span<const unsigned char> input, bool pad = true);
+ * @param[in] in        The socket address string to split.
+ * @param[out] portOut  Port-portion of the input, if found and parsable.
+ * @param[out] hostOut  Host-portion of the input, if found.
+ * @return              true if port-portion is found and within its allowed range, otherwise false
```
```suggestion
 * @return              true if port-portion is absent or within its allowed range, otherwise false
```

luke-jr commented on src/util/strencodings.cpp:
```diff
@@ -107,25 +107,35 @@ std::vector<unsigned char> ParseHex(const std::string& str)
+          valid = true;
         }
+    } else {
+      valid = true;
```
IMO we should be setting `portOut` to `0` here explicitly. Maybe make it a bugfix commit before your others.

amadeuszpawlik replied:
The way this has been used up to now makes for neat code:
https://github.com/bitcoin/bitcoin/blob/25290071c434638e2719a99784572deef44542ad/src/netbase.cpp#L203-L205
Don't you think it's a bit neater this way?

luke-jr replied:
Yes, I agree.

luke-jr commented on src/init.cpp:
```diff
@@ -1276,6 +1276,43 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+        for (std::string socketAddr : args.GetArgs(port_option)) {
+            std::string hostOut;
+            uint16_t portOut{0};
+            if (port_option == "-bind") {
+                socketAddr = socketAddr.substr(0, socketAddr.rfind('='));
+            }
```
Hmm, maybe better to keep `-bind` separate for this.

luke-jr commented on src/util/error.cpp:
```diff
@@ -45,6 +45,11 @@ bilingual_str ResolveErrMsg(const std::string& optname, const std::string& strBi
     return strprintf(_("Cannot resolve -%s address: '%s'"), optname, strBind);
 }
 
+bilingual_str InvalidPortErrMsg(const std::string& optname, const std::string& strPort)
```
nit: Just call it `port`, or maybe `port_str`; we don't use Hungarian variable names in new code.

luke-jr replied:
Actually, something like `invalid_value` might be better, since this *isn't* just a bare port number.

amadeuszpawlik replied:
Thanks again for your input, Luke

luke-jr commented on src/init.cpp:
```diff
@@ -1276,6 +1276,48 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+        "-zmqpubrawblock",
+        "-zmqpubsequence",
+    }) {
+        for (std::string& socket_addr : args.GetArgs(port_option)) {
```
nit

```suggestion
        for (const std::string& socket_addr : args.GetArgs(port_option)) {
```

luke-jr commented on src/init.cpp:
```diff
@@ -1276,6 +1276,48 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+        }
+    }
+
+    for (std::string& socket_addr : args.GetArgs("-bind")) {
```
nit

```suggestion
    for (const std::string& socket_addr : args.GetArgs("-bind")) {
```

luke-jr commented on src/init.cpp:
```diff
@@ -1276,6 +1276,48 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+        uint16_t port_out{0};
+        std::string bind_socket_addr = socket_addr.substr(0, socket_addr.rfind('='));
+        if (!SplitHostPort(bind_socket_addr, port_out, host_out)) {
+            return InitError(InvalidPortErrMsg("-bind", bind_socket_addr));
```
Might be better to include the full value in the error? idk

```suggestion
            return InitError(InvalidPortErrMsg("-bind", socket_addr));
```

luke-jr reviewed (approved):
utACK as-is, a few nits

prayank23 commented:
I was fuzzing `-port` with different wordlists and noticed something weird for 2 values in [`big-list-of-naughty-strings.txt`](https://raw.githubusercontent.com/danielmiessler/SecLists/master/Fuzzing/big-list-of-naughty-strings.txt):

It uses port as 1 when `-port=1/2` or `-port=1E2`. Is this expected?

Can directly try those values or go through each using:

```
wget https://raw.githubusercontent.com/danielmiessler/SecLists/master/Fuzzing/big-list-of-naughty-strings.txt
while read line;do bitcoind -port="$line";done<big-list-of-naughty-strings.txt
```

amadeuszpawlik commented:
Thanks for catching this @pryank23, changed the way `-port` and `-rpcport` are parsed.
`-port=1/2` and the like are caught and not defaulting to 1 anymore, and the error returns the exact string that has been passed as argument.

prayank23 reviewed (approved):
tACK https://github.com/bitcoin/bitcoin/pull/22087/commits/51155b512a7acdc6950b71284c7711386c6f4a0c

I could not find anymore issues however I am not confident and think this PR needs more fuzzing and review.

luke-jr commented:
```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

re-utACK 51155b512a7acdc6950b71284c7711386c6f4a0c
-----BEGIN PGP SIGNATURE-----

iQQzBAEBCAAdFiEE5GOpP18xF+7ebHMWvQKUJCH0iJ8FAmIn8TkACgkQvQKUJCH0
iJ9taCAAoTF45MLzSW5rL15SNoJcy+/H1z2QLKrpW6NzG5ktgY5aJNCJu+BR0bVQ
ZIit/5eKECGf2mXf+ZBpknGU/O4QZZHHe4qKC8A7AhWFVU0pgyDRE3U9Yir23Cev
8oRb8fji21e5p2JuL/bU1WTUMla0B3YBmPt8DyzFe35k3DyOGiriDkh+dut8eWUn
LkCe/JcVf8VgjpLAL2ZDGp9Drhu+V6y/JPMiFQ7IYdo5H70JErKjd9I6rCfaYg37
jaMVNgK0yZ8G5SDBdhjITb4d7QK700/ZbHfqBRU3yXabGE4fJoRRkdnmbcMG4MV0
iC/ZGb9yhhSFEK/rkJFqYca2png3In2gl/NsDklUG9+mYX3SwEyKat4oXc5n3d6L
fjYezT9rXH6mjKWtnBvRXVfkE0X3bcbv70jI8fQKwCpHPbl9Y8mTlQPl7iZS+93c
50zUUSWV4fgI5FTVbm9lAiCrXNEa8vQJzi5ZnSzKS5BDcQXdGFq3/sgfpDkTAg9f
kA5JZtoB11t7nNObFZJVc5oN2QR4e6Hjs12ZguqAQQfZidXvLb7GPzU4ttjaod2y
B+/PInkac+HiDfZ7EkeYTShB3WO/CxjISXGWDun/vUi0v2LfI8pI8oCOzMhNoXw+
yPq5Sb3xxXJXM3k0TSyv1HWtLCnEWvm9GdEJe+Zrqr7ebcH/r1nRurPpXP8oA3et
PqWeikv9r/s9FTAXjU766wmomivlKDqV2RAn9TT0unrKgMMQlC1UXp02jL9FoABX
6hsy/JY9HUQ41jY4fR2Oo4R5uuPfpOAkhffFDe30RZWoRIq5+zsStARsy9SSoeVt
fxmzOV//pRELFY3VsQ6x3EFAj5yT2paXDadst2i0Aa40obI//iA42Qz7up7MAid6
rB3ukbme8NnjDZ0Njc4aNGB6xdH/Cm71jvgA4bpDXWt5XQtLsyKk/EvzjIgnf6/q
D0Iete9WYYGefYf3K+Nd/O7xyYwY64eeAveeSHJrTiwt2ht0DcLIasRiOpvYJzCu
QfS/vzVMoZ0FyYBi7+ZRuckBLVBhmN86HuNb6zZy2BZIOJKcTv7pqTL7H77evygx
R+DPgYWUBWOQfYnS7FyOqp/RQ5mQjiylxhtQdhDmhe00ZWthWcflwp9zNSNsnUxS
jVjlMhZvBx/ehiOZzBpsOX3K9pY9zqAOLmMHaLrG0r0krMtwxBZ3otI9DrwNAvhw
XxcR6Cu65b/8GUBB74/R864zhIi8xreUirkwjxXbA1byjgpppajyw/basoShsdlw
fATGwITm4DjaURrvms4/ox4wFsO2IzwMEXTIL9Qb76eEI4sZzgbO+h3/IlxTbdSu
sl2aM0INVY7FQ/2GHBcXKALHdglOrA==
=EKhU
-----END PGP SIGNATURE-----
```

luke-jr reviewed (changes requested):
Sorry, GCC is whining about the loops in init. Probably should make it happy.

```
init.cpp: In function ‘bool AppInitMain(NodeContext&, interfaces::BlockAndHeaderTipInfo*)’:
init.cpp:1280:29: warning: loop variable ‘port_option’ of type ‘const string&’ {aka ‘const std::__cxx11::basic_string<char>&’} binds to a temporary constructed from type ‘const char* const’ [-Wrange-loop-construct]
 1280 |     for (const std::string& port_option : {
      |                             ^~~~~~~~~~~
init.cpp:1280:29: note: use non-reference type ‘const string’ {aka ‘const std::__cxx11::basic_string<char>’} to make the copy explicit or ‘const char* const&’ to prevent copying
init.cpp:1293:29: warning: loop variable ‘port_option’ of type ‘const string&’ {aka ‘const std::__cxx11::basic_string<char>&’} binds to a temporary constructed from type ‘const char* const’ [-Wrange-loop-construct]
 1293 |     for (const std::string& port_option : {
      |                             ^~~~~~~~~~~
init.cpp:1293:29: note: use non-reference type ‘const string’ {aka ‘const std::__cxx11::basic_string<char>’} to make the copy explicit or ‘const char* const&’ to prevent copying
```

luke-jr commented:
Looking at the options available, I think it'd be best to just drop the `&` in the loop for now (option A).

```c++
    for (const std::string port_option : {
        "-port",
        "-rpcport",
    }) {
```

Option B: Explicitly declare the type of the list. But I think this is slower than option A for no benefit.

```c++
    for (const std::string& port_option : std::vector<std::string>{
        "-port",
        "-rpcport",
    }) {
```

Option C: Use `std::string_literals`

```c++
    using namespace std::string_literals;
    for (const std::string& port_option : {
        "-port"s,
        "-rpcport"s,
    }) {
```

Option D: Use `std::array<std::string>`? Kinda ugly to have to specify the number of items though IMO.

```c++
    for (const std::string& port_option : (const std::array<const std::string, 2>){
        "-port",
        "-rpcport",
    }) {
```

Option E: C++20 supposedly makes `std::vector` more flexible, so it could be a compile-time constexpr. But hopefully this gets merged before C++20 is required :)

amadeuszpawlik commented:
Thanks @luke-jr , done!
Rebased & fixed failing [feature_proxy.py](https://github.com/bitcoin/bitcoin/pull/22087/files#diff-156fee7ba1e072ea2ff94156a03c2c8b5e4885ee2181ecff97d3c772723802fb)-tests

luke-jr commented on test/functional/feature_proxy.py:
```diff
@@ -320,19 +320,34 @@ def networks_dict(d):
-        self.nodes[1].extra_args = ["-proxy=abc:def"]
-        msg = "Error: Invalid -proxy address or hostname: 'abc:def'"
+        self.log.info("Test passing invalid -proxy hostname raises expected init error")
+        self.nodes[1].extra_args = ["-proxy=abc:23456"]
```
This now performs and depends on the outcome of a DNS lookup, I think. Need to ensure it always fails.

amadeuszpawlik replied:
Good catch. As per the RFC 2181: 

> The DNS itself places only one restriction on the particular labels
   that can be used to identify resource records [...] The length of
   any one label is limited to between 1 and 63 octets.

I therefore propose to set the hostname with an invalid 0-length label (e.g. `abc..abc`), which should always fail.

luke-jr reviewed (approved):
```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA256

re-utACK 1b6f8d3ea0876ede1fff97d725d2094f9e1dac68
-----BEGIN PGP SIGNATURE-----

iQQzBAEBCAAdFiEE5GOpP18xF+7ebHMWvQKUJCH0iJ8FAmJXETQACgkQvQKUJCH0
iJ8dfh/+JdJ73VBn/G/wmten/hIhXMMKZFGTAdC2cr/xVFq0cdst6KAfEf25J5YM
6q0iadBHV8My/3BRTQPxpPxQMAD9pAwaol17yGxWpwQxD+Y9DVSfYjTYRwdJOTsw
ClKKcNu2nbBr6iNuE5XI0i9Y/D2ycx86cH9hDgKUwRh2BMFivh89awZ9I2zAP8UE
+uWM3Kst5EC5cegkxIWj03idbrqHHPixq51Ium4Fr+XXokSbz9QSTNj6snwUujHr
WROCqiNDB1x1BhZ4F48GRTh5IGTBMhEuTWPucKRxGS89064YxdV4UsQna7H8eJ5o
kqmMeBcOTib70nsEirPPrFVeJUfrb6/XvD4TN5+8E8PBn9O3WkFLwnjzLbSscMgl
uELth1n1vNFMu9UvKWwgxJJXBULvN+uFAaQzkbr0+L3GjiRKaCuYJNJBCojN15V4
lRdALdZX6pgyvbYYwlQhKtbl69fGXAoO1ZSI3CRzrejfmqDOxI7E7w6rP/TN2LIC
ifHb8q6u3Jza0B334bbZMBHwbFHIgC3TfOPjf1JJBgZNIq2l//KuqzN4xu8ZcTyJ
A4UAH4JaWQoB1krU1HF112bIMPwon5ZrMiKZU4P0a3mOuSlwL294/z1rEjg8wUhe
mHg9RdOVxdFGyQ/2ALc7SgpBOcQilaheQ2ARUw9osjSDqpCJYpHEhJ60L05TWDvp
7mn61pdiOc0Hq3YUKNPWzRkdslo5Lx/x0pjHZwMm2wRsz202SUquWosC0NhwMRX7
vkINH/UFzKsG+k/mTUCx0OJSzmJGLSXRtE+ByozJGK3eEzvUx+sapJpsIHvBT98D
S66neNcVZiS4pqR08MwtddcSs47pyNkWn6YXHAP7LOrkkD1gu2XRntiLLLETh4k8
/8HrQrS3teIVVpCazcDCYIucH3Bd5n9LU8VWMX7a1Lw5ts6+mhQ06Qt/zKPApQ6E
DrB5lQoeXzOIH/bUvXDF79WFU6yVyww28u8U4N65KUyVdh42fddRQiFGvvvcY9m4
/Vq1hMUDOZHAkBhKMQ5rpoRa05jFQPHbaae70+6cvEGrMwDO6V7ON4avVtATOvpa
S/DIFpDOpd0jKf/5l9TM6pm8FZAWK7vf2Qr8QkCL9gfFSnqQZiY2Bh8IYDpNCv0p
c88Kfp6IQDJOK8XmtQdHPPYRC74QST+fIXp0QIwK2Qrf8lq/lbxxGY9YqR3lnz7f
M6rKMbD2SxBs6J5H/EvYHb8PnKIalfCfnTQa4zYo7ogFirYxJc1IEe4gfTN2GoNm
BwAwiVD0I5VLXXhGPuQcyWE6wlvl/RG4t7Xuin/9aAheel5+KbWB7Hk78FGLNmC/
Lv0trYwADpAr2JFsrqaSXYQ/dydTQA==
=oWrb
-----END PGP SIGNATURE-----
```

luke-jr commented on src/init.cpp:
```diff
@@ -1290,6 +1290,50 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+        "-whitebind",
+        "-zmqpubhashblock",
+        "-zmqpubhashtx",
+        "-zmqpubrawblock",
```
`"-zmqpubrawtx",` is missing ?

amadeuszpawlik replied:
resolved! thanks!

ryanofsky commented on src/init.cpp:
```diff
@@ -1290,6 +1290,51 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+        "-rpcport",
+    }) {
+        if (args.IsArgSet(port_option)) {
+            const std::string port = args.GetArgs(port_option)[0];
```
In commit "Validate `port` options" (1dae86bfd2259e4d9a9f1cd292a10cc17f626876)

Suggest replacing `args.GetArgs(port_option)[0]` with `args.GetArg(port_option)`. `IsArgSet` will true and the `[0]` subscript will be out of bounds if the argument is negated (`-noport`)

amadeuszpawlik replied:
Good catch, thanks

ryanofsky reviewed (commented):
Reviewed 1dae86bfd2259e4d9a9f1cd292a10cc17f626876, and it looks almost ok, but I am concerned different `SplitHostPort` `hostOut` values when port is invalid could potentially be bad for callers. Or even if it doesn't break anything, it seems better just to add a additional return value not change `hostOut` values in this PR.

It would also be good to note the change of behavior in the PR description or release notes that `-port` and `-rpcport` values that previously worked and were considered valid can now result in errors. For example port values with leading whitespace or trailing non-digit characters.

I do think code in second commit is a little awkward because it is parsing and validating address and port options in different place where they are used. But probably this is fine, since in the future this will probably be cleaned up by introducing options structs and moving new code here to a function like [`ApplyArgsManOptions`](https://github.com/bitcoin/bitcoin/blob/9fcdb9f3a044330d3d7515fa35709102c98534d2/src/node/mempool_args.cpp#L41) that does parsing and error-checking all in one place.

fanquake commented:
@amadeuszpawlik any chance you're still around / interested in responding to the latest review feedback?

amadeuszpawlik commented:
> @amadeuszpawlik any chance you're still around / interested in responding to the latest review feedback?

Yes, definitely. Will have a look at it shortly 👍

amadeuszpawlik commented on src/util/strencodings.cpp:
```diff
@@ -95,25 +95,33 @@ std::vector<unsigned char> ParseHex(std::string_view str)
-        if (ParseUInt16(in.substr(colon + 1), &n)) {
-            in = in.substr(0, colon);
-            portOut = n;
+        if (!in.substr(colon + 1).empty()) {
```
Agreed

luke-jr commented on src/util/strencodings.cpp:
```diff
@@ -95,25 +95,31 @@ std::vector<unsigned char> ParseHex(std::string_view str)
         uint16_t n;
-        if (ParseUInt16(in.substr(colon + 1), &n)) {
-            in = in.substr(0, colon);
+        if (ParseUInt16(in.substr(colon+1), &n)) {
```
Why are you changing this? It's better with the spaces IMO

amadeuszpawlik commented on src/util/strencodings.cpp:
```diff
@@ -95,25 +95,33 @@ std::vector<unsigned char> ParseHex(std::string_view str)
+              valid = (portOut != 0);
+            }
         }
+        in = in.substr(0, colon);
```
I agree, this PR is about validating port-options and not cleaning up the host string.

amadeuszpawlik commented:
Thanks @ryanofsky for taking your time. I have now resolved all your comments and added a release-note.

ryanofsky reviewed (approved):
Code review ACK 04526787b5f6613d1f1ad78434e1dd24ab88dd76. Just suggested changes since last review: reverting some SplitHostPort changes, adding release notes, avoiding 'GetArgs[0]` problem.

luke-jr commented:
utACK 04526787b5f6613d1f1ad78434e1dd24ab88dd76

fanquake commented:
@ryanofsky just fyi. You've re-ACK the previous commit hash in your latest comment

ryanofsky commented:
> @ryanofsky just fyi. You've re-ACK the previous commit hash in your latest comment

Thank you, fixed now

JayBitron commented:
@fanquake We now have new error: Error: Invalid port specified in -zmqpubhashblock: 'ipc:///root/test/zmq.block.hash'
Error: Invalid port specified in -zmqpubhashblock: 'ipc:///root/test/zmq.block.hash'
